opm/simulators/WellSwitchingLogger.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+  Copyright 2016 Dr. Blatt - HPC-Simulation-Software & Services
+  Copyright 2016 Statoil AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif // HAVE_CONFIG_H
+
+#include <opm/simulators/WellSwitchingLogger.hpp>
+
+namespace Opm
+{
+namespace wellhelpers
+{
+
+#if HAVE_MPI
+int WellSwitchingLogger::calculateMessageSize(std::vector<int>& well_name_length)
+{
+
+    // Each process will send a message to the root process with
+    // the following data:
+    // total number of switches, for each switch the length of the
+    // well name, for each switch the well name and the two controls.
+    well_name_length.reserve(switchMap_.size());
+
+    for(const auto& switchEntry : switchMap_)
+    {
+        int length = switchEntry.first.size() +1;  //we write an additional \0
+        well_name_length.push_back(length);
+    }
+
+    // compute the message size
+    int message_size = 0;
+    int increment = 0;
+    // number of switches
+    MPI_Pack_size(1, MPI_INT, MPI_COMM_WORLD, &message_size);
+    message_size += increment;
+    // const char* length include delimiter for each switch
+    MPI_Pack_size(switchMap_.size(), MPI_INT, MPI_COMM_WORLD, &increment);
+    message_size += increment;
+    // for each well the name + two controls in one write
+    auto length = well_name_length.begin();
```

Comment:
I think it is an important distinction: is `++length` increasing a quantity or moving an iterator, for example? Pointers and references are different since they still semantically represent just a single object, but an iterator is not and therefore needs a name that says so.